Resteasy Reactive/Panache requests reactive Hibernate session from executor thread if request is large

### Describe the bug

I have a simple project which accepts some data via JSON through RESTEasy Reactive and saves it to the database using Hibernate Reactive with Panache. This works fine as long as the request object is small. However, if the request object is large (in my case, a few megabytes), then the request fails with `HR000068: This method should exclusively be invoked from a Vert.x EventLoop thread; currently running on thread 'executor-thread-0'` (see below for the full stack trace).


### Expected behavior

The request should always finish successfully, regardless of the size of the request object.

### Actual behavior

The request fails and returns a 500 status code. The following logging output is generated:
```
2023-03-04 16:57:10,768 ERROR [org.jbo.res.rea.ser.cor.RuntimeExceptionMapper] (executor-thread-0) An operation that needed be run on a Vert.x EventLoop thread was run on a worker pool thread. This likely means you need to annotate de.markushimmel.Resource#create(class de.markushimmel.RequestObject) with @io.smallrye.common.annotation.NonBlocking. Alternatively you can annotate the class de.markushimmel.Resource to make every method on the class run on a Vert.x EventLoop thread, or annotate your sub class of the jakarta.ws.rs.core.Application class to affect the entire application
2023-03-04 16:57:10,774 ERROR [io.qua.ver.htt.run.QuarkusErrorHandler] (executor-thread-0) HTTP Request to /resource/create failed, error id: bf8e7471-cb50-4817-aeaf-eb549c080252-1: java.lang.IllegalStateException: HR000068: This method should exclusively be invoked from a Vert.x EventLoop thread; currently running on thread 'executor-thread-0'
        at org.hibernate.reactive.common.InternalStateAssertions.assertUseOnEventLoop(InternalStateAssertions.java:40)
        at org.hibernate.reactive.mutiny.impl.MutinySessionFactoryImpl.connection(MutinySessionFactoryImpl.java:177)
        at org.hibernate.reactive.mutiny.impl.MutinySessionFactoryImpl.lambda$openSession$0(MutinySessionFactoryImpl.java:111)
        at io.smallrye.context.impl.wrappers.SlowContextualSupplier.get(SlowContextualSupplier.java:21)
        at io.smallrye.mutiny.operators.uni.builders.UniCreateFromCompletionStage.subscribe(UniCreateFromCompletionStage.java:24)
        at io.smallrye.mutiny.operators.AbstractUni.subscribe(AbstractUni.java:36)
        at io.smallrye.mutiny.operators.uni.UniRunSubscribeOn.lambda$subscribe$0(UniRunSubscribeOn.java:27)
        at org.hibernate.reactive.context.impl.VertxContext.execute(VertxContext.java:90)
        at io.smallrye.mutiny.operators.uni.UniRunSubscribeOn.subscribe(UniRunSubscribeOn.java:25)
        at io.smallrye.mutiny.operators.AbstractUni.subscribe(AbstractUni.java:36)
        at io.smallrye.mutiny.operators.uni.UniOnItemTransformToUni.subscribe(UniOnItemTransformToUni.java:25)
        at io.smallrye.mutiny.operators.AbstractUni.subscribe(AbstractUni.java:36)
        at io.smallrye.mutiny.operators.uni.UniOnItemTransform.subscribe(UniOnItemTransform.java:22)
        at io.smallrye.mutiny.operators.AbstractUni.subscribe(AbstractUni.java:36)
        at io.smallrye.mutiny.operators.uni.UniOnItemConsume.subscribe(UniOnItemConsume.java:30)
        at io.smallrye.mutiny.operators.AbstractUni.subscribe(AbstractUni.java:36)
        at io.smallrye.mutiny.operators.uni.UniOnItemTransformToUni.subscribe(UniOnItemTransformToUni.java:25)
        at io.smallrye.mutiny.operators.AbstractUni.subscribe(AbstractUni.java:36)
        at io.smallrye.mutiny.operators.uni.UniOnItemOrFailureFlatMap.subscribe(UniOnItemOrFailureFlatMap.java:27)
        at io.smallrye.mutiny.operators.AbstractUni.subscribe(AbstractUni.java:36)
        at io.smallrye.mutiny.operators.uni.UniOnItemTransform.subscribe(UniOnItemTransform.java:22)
        at io.smallrye.mutiny.operators.AbstractUni.subscribe(AbstractUni.java:36)
        at io.smallrye.mutiny.operators.uni.UniOnItemOrFailureFlatMap.subscribe(UniOnItemOrFailureFlatMap.java:27)
        at io.smallrye.mutiny.operators.AbstractUni.subscribe(AbstractUni.java:36)
        at io.smallrye.mutiny.groups.UniSubscribe.withSubscriber(UniSubscribe.java:51)
        at io.smallrye.mutiny.groups.UniSubscribe.with(UniSubscribe.java:110)
        at io.smallrye.mutiny.groups.UniSubscribe.with(UniSubscribe.java:88)
        at org.jboss.resteasy.reactive.server.handlers.UniResponseHandler.handle(UniResponseHandler.java:19)
        at io.quarkus.resteasy.reactive.server.runtime.QuarkusResteasyReactiveRequestContext.invokeHandler(QuarkusResteasyReactiveRequestContext.java:123)
        at org.jboss.resteasy.reactive.common.core.AbstractResteasyReactiveContext.run(AbstractResteasyReactiveContext.java:145)
        at io.quarkus.vertx.core.runtime.VertxCoreRecorder$14.runWith(VertxCoreRecorder.java:576)
        at org.jboss.threads.EnhancedQueueExecutor$Task.run(EnhancedQueueExecutor.java:2449)
        at org.jboss.threads.EnhancedQueueExecutor$ThreadBody.run(EnhancedQueueExecutor.java:1478)
        at org.jboss.threads.DelegatingRunnable.run(DelegatingRunnable.java:29)
        at org.jboss.threads.ThreadLocalResettingRunnable.run(ThreadLocalResettingRunnable.java:29)
        at io.netty.util.concurrent.FastThreadLocalRunnable.run(FastThreadLocalRunnable.java:30)
        at java.base/java.lang.Thread.run(Thread.java:1589)

2023-03-04 16:57:10,777 ERROR [org.jbo.res.rea.com.cor.AbstractResteasyReactiveContext] (executor-thread-0) Request failed: java.lang.IllegalStateException: HR000068: This method should exclusively be invoked from a Vert.x EventLoop thread; currently running on thread 'executor-thread-0'
        at org.hibernate.reactive.common.InternalStateAssertions.assertUseOnEventLoop(InternalStateAssertions.java:40)
        at org.hibernate.reactive.mutiny.impl.MutinySessionFactoryImpl.connection(MutinySessionFactoryImpl.java:177)
        at org.hibernate.reactive.mutiny.impl.MutinySessionFactoryImpl.lambda$openSession$0(MutinySessionFactoryImpl.java:111)
        at io.smallrye.context.impl.wrappers.SlowContextualSupplier.get(SlowContextualSupplier.java:21)
        at io.smallrye.mutiny.operators.uni.builders.UniCreateFromCompletionStage.subscribe(UniCreateFromCompletionStage.java:24)
        at io.smallrye.mutiny.operators.AbstractUni.subscribe(AbstractUni.java:36)
        at io.smallrye.mutiny.operators.uni.UniRunSubscribeOn.lambda$subscribe$0(UniRunSubscribeOn.java:27)
        at org.hibernate.reactive.context.impl.VertxContext.execute(VertxContext.java:90)
        at io.smallrye.mutiny.operators.uni.UniRunSubscribeOn.subscribe(UniRunSubscribeOn.java:25)
        at io.smallrye.mutiny.operators.AbstractUni.subscribe(AbstractUni.java:36)
        at io.smallrye.mutiny.operators.uni.UniOnItemTransformToUni.subscribe(UniOnItemTransformToUni.java:25)
        at io.smallrye.mutiny.operators.AbstractUni.subscribe(AbstractUni.java:36)
        at io.smallrye.mutiny.operators.uni.UniOnItemTransform.subscribe(UniOnItemTransform.java:22)
        at io.smallrye.mutiny.operators.AbstractUni.subscribe(AbstractUni.java:36)
        at io.smallrye.mutiny.operators.uni.UniOnItemConsume.subscribe(UniOnItemConsume.java:30)
        at io.smallrye.mutiny.operators.AbstractUni.subscribe(AbstractUni.java:36)
        at io.smallrye.mutiny.operators.uni.UniOnItemTransformToUni.subscribe(UniOnItemTransformToUni.java:25)
        at io.smallrye.mutiny.operators.AbstractUni.subscribe(AbstractUni.java:36)
        at io.smallrye.mutiny.operators.uni.UniOnItemOrFailureFlatMap.subscribe(UniOnItemOrFailureFlatMap.java:27)
        at io.smallrye.mutiny.operators.AbstractUni.subscribe(AbstractUni.java:36)
        at io.smallrye.mutiny.operators.uni.UniOnItemTransform.subscribe(UniOnItemTransform.java:22)
        at io.smallrye.mutiny.operators.AbstractUni.subscribe(AbstractUni.java:36)
        at io.smallrye.mutiny.operators.uni.UniOnItemOrFailureFlatMap.subscribe(UniOnItemOrFailureFlatMap.java:27)
        at io.smallrye.mutiny.operators.AbstractUni.subscribe(AbstractUni.java:36)
        at io.smallrye.mutiny.groups.UniSubscribe.withSubscriber(UniSubscribe.java:51)
        at io.smallrye.mutiny.groups.UniSubscribe.with(UniSubscribe.java:110)
        at io.smallrye.mutiny.groups.UniSubscribe.with(UniSubscribe.java:88)
        at org.jboss.resteasy.reactive.server.handlers.UniResponseHandler.handle(UniResponseHandler.java:19)
        at io.quarkus.resteasy.reactive.server.runtime.QuarkusResteasyReactiveRequestContext.invokeHandler(QuarkusResteasyReactiveRequestContext.java:123)
        at org.jboss.resteasy.reactive.common.core.AbstractResteasyReactiveContext.run(AbstractResteasyReactiveContext.java:145)
        at io.quarkus.vertx.core.runtime.VertxCoreRecorder$14.runWith(VertxCoreRecorder.java:576)
        at org.jboss.threads.EnhancedQueueExecutor$Task.run(EnhancedQueueExecutor.java:2449)
        at org.jboss.threads.EnhancedQueueExecutor$ThreadBody.run(EnhancedQueueExecutor.java:1478)
        at org.jboss.threads.DelegatingRunnable.run(DelegatingRunnable.java:29)
        at org.jboss.threads.ThreadLocalResettingRunnable.run(ThreadLocalResettingRunnable.java:29)
        at io.netty.util.concurrent.FastThreadLocalRunnable.run(FastThreadLocalRunnable.java:30)
        at java.base/java.lang.Thread.run(Thread.java:1589)
```

### How to Reproduce?

Reproducer: https://github.com/TwoFX/resteasy-reactive-large-request-bug

Steps to reproduce:
Run `./mvnw verify`. There are two tests, one for a request body of around 5000 bytes and one for a request body of around 5000000 bytes. The first test succeeds, and the second test fails.

### Output of `uname -a` or `ver`

Linux markus-laptop 6.2.1-arch1-1 #1 SMP PREEMPT_DYNAMIC Sun, 26 Feb 2023 03:39:23 +0000 x86_64 GNU/Linux

### Output of `java -version`

```
openjdk version "19.0.2" 2023-01-17
OpenJDK Runtime Environment (build 19.0.2+7)
OpenJDK 64-Bit Server VM (build 19.0.2+7, mixed mode)
```

### GraalVM version (if different from Java)

_No response_

### Quarkus version or git rev

3.0.0.Alpha4

### Build tool (ie. output of `mvnw --version` or `gradlew --version`)

```
Apache Maven 3.8.6 (84538c9988a25aec085021c365c560670ad80f63)
Maven home: /home/markus/.m2/wrapper/dists/apache-maven-3.8.6-bin/67568434/apache-maven-3.8.6
Java version: 19.0.2, vendor: N/A, runtime: /usr/lib/jvm/java-19-openjdk
Default locale: en_US, platform encoding: UTF-8
OS name: "linux", version: "6.2.1-arch1-1", arch: "amd64", family: "unix"
```

### Additional information

I have verified that the problem is NOT present in Quarkus 2.16.4.Final.

I tried to test the current main, but unfortunately after changing the group id and the version the reproducer fails to build. Maven gives the following error:
```
[WARNING] The POM for io.quarkus:quarkus-hibernate-reactive-panache:jar:999-20230217.022801-642 is invalid, transitive dependencies (if any) will not be available: 1 problem was encountered while building the effective model for io.quarkus:quarkus-hibernate-reactive-panache:999-SNAPSHOT
[ERROR] 'dependencies.dependency.version' for org.hibernate:hibernate-jpamodelgen-jakarta:jar is missing. @ 
```
I'm not sure what I'm doing wrong here. It would be very nice if someone could assist me with this, as this is currently preventing me from testing potential fixes.

I have digged into the bug a bit. The reason why we're on an executor thread in the first place is that the `InputHandler` does this if the request object is too large. Thus, for small request objects, we enter the `UniResultHandler` on a Vertx event loop thread, while for large request objects, we enter it on an executor thread.

Now, (but, as noted above, I haven't tested any of this), my guess is that the first commit with this problem is 47b10aed49150aa5e71d7eae9e79b69e6787a0f9. Since then, the Uni subscription created by [`Panache.withTransaction`](https://github.com/quarkusio/quarkus/commit/47b10aed49150aa5e71d7eae9e79b69e6787a0f9#diff-66b7f197f4be0aa040828ae4fc8e1f05b72272ed4322661c657267203ad96101) uses a different `Executor`. In Quarkus 2, it is `AbstractJpaOperations::executeInVertxEventLoop` (see [here](https://github.com/quarkusio/quarkus/blob/2.16/extensions/panache/hibernate-reactive-panache-common/runtime/src/main/java/io/quarkus/hibernate/reactive/panache/common/runtime/AbstractJpaOperations.java#L138)). In Quarkus 3, it is the `org.hibernate.reactive.context.impl.VertxContext` (see [here](https://github.com/hibernate/hibernate-reactive/blob/7a7565bd81f3ba2b87af5022054fc358434c4457/hibernate-reactive-core/src/main/java/org/hibernate/reactive/mutiny/impl/MutinySessionFactoryImpl.java#L71)).

Now, it turns out that [`VertxContext::execute`](https://github.com/hibernate/hibernate-reactive/blob/7a7565bd81f3ba2b87af5022054fc358434c4457/hibernate-reactive-core/src/main/java/org/hibernate/reactive/context/impl/VertxContext.java#L76) doesn't actually always execute the runnable on a Vertx event loop thread, as you can see in the `else` branch. I don't understand any of this well enough to know whether this is intentional or not -- perhaps, the line `runnable.run()` (line 90) should actually be `currentContext.executeInContext( x -> runnable.run() )`. So it is actually possible that this isn't a problem with Quarkus, but with Hibernate Reactive. Again, as I mentioned, I haven't tried this potential fix because I can't get Quarkus to build properly locally.

I would very much like to contribute a fix for this myself if possible, I only need some feedback if my analysis is correct and some assistance with the "POM is invalid" problem described above.

--- a/extensions/resteasy-reactive/quarkus-resteasy-reactive-jackson/deployment/src/test/java/io/quarkus/resteasy/reactive/jackson/deployment/test/SimpleJsonResource.java
+++ b/extensions/resteasy-reactive/quarkus-resteasy-reactive-jackson/deployment/src/test/java/io/quarkus/resteasy/reactive/jackson/deployment/test/SimpleJsonResource.java
@@ -202,9 +202,8 @@ public class SimpleJsonResource extends SuperClass<Person> {
     @Produces(MediaType.APPLICATION_JSON)
     @Consumes(MediaType.APPLICATION_JSON)
     public Person personTest(Person person) {
-        //large requests should get bumped from the IO thread
-        if (!BlockingOperationControl.isBlockingAllowed()) {
-            throw new RuntimeException("should have dispatched");
+        if (BlockingOperationControl.isBlockingAllowed()) {
+            throw new RuntimeException("should have dispatched back to event loop");
         }
         return person;
     }

--- a/extensions/resteasy-reactive/quarkus-resteasy-reactive-jaxb/deployment/src/test/java/io/quarkus/resteasy/reactive/jaxb/deployment/test/SimpleXmlTest.java
+++ b/extensions/resteasy-reactive/quarkus-resteasy-reactive-jaxb/deployment/src/test/java/io/quarkus/resteasy/reactive/jaxb/deployment/test/SimpleXmlTest.java
@@ -262,9 +262,8 @@ public class SimpleXmlTest {
         @Produces(MediaType.APPLICATION_XML)
         @Consumes(MediaType.APPLICATION_XML)
         public Person personTest(Person person) {
-            //large requests should get bumped from the IO thread
-            if (!BlockingOperationControl.isBlockingAllowed()) {
-                throw new RuntimeException("should have dispatched");
+            if (BlockingOperationControl.isBlockingAllowed()) {
+                throw new RuntimeException("should have dispatched back to event loop");
             }
             return person;
         }

--- a/extensions/resteasy-reactive/quarkus-resteasy-reactive-jsonb/deployment/src/test/java/io/quarkus/resteasy/reactive/jsonb/deployment/test/SimpleJsonResource.java
+++ b/extensions/resteasy-reactive/quarkus-resteasy-reactive-jsonb/deployment/src/test/java/io/quarkus/resteasy/reactive/jsonb/deployment/test/SimpleJsonResource.java
@@ -104,8 +104,8 @@ public class SimpleJsonResource extends SuperClass<Person> {
     @Consumes(MediaType.APPLICATION_JSON)
     public Person personTest(Person person) {
         //large requests should get bumped from the IO thread
-        if (!BlockingOperationControl.isBlockingAllowed()) {
-            throw new RuntimeException("should have dispatched");
+        if (BlockingOperationControl.isBlockingAllowed()) {
+            throw new RuntimeException("should have dispatched back to event loop");
         }
         return person;
     }

--- a/independent-projects/resteasy-reactive/server/runtime/src/main/java/org/jboss/resteasy/reactive/server/core/startup/RuntimeResourceDeployment.java
+++ b/independent-projects/resteasy-reactive/server/runtime/src/main/java/org/jboss/resteasy/reactive/server/core/startup/RuntimeResourceDeployment.java
@@ -286,6 +286,7 @@ public class RuntimeResourceDeployment {
         }
         // form params can be everywhere (field, beanparam, param)
         boolean checkWithFormReadRequestFilters = false;
+        boolean inputHandlerEngaged = false;
         if (method.isFormParamRequired() || hasWithFormReadRequestFilters) {
             // read the body as multipart in one go
             handlers.add(new FormBodyHandler(bodyParameter != null, executorSupplier, method.getFileFormNames()));
@@ -296,6 +297,7 @@ public class RuntimeResourceDeployment {
                     // allow the body to be read by chunks
                     handlers.add(new InputHandler(resteasyReactiveConfig.getInputBufferSize(), executorSupplier));
                     checkWithFormReadRequestFilters = true;
+                    inputHandlerEngaged = true;
                 }
             }
         }
@@ -324,6 +326,10 @@ public class RuntimeResourceDeployment {
             }
             handlers.add(new RequestDeserializeHandler(typeClass, genericType, consumesMediaTypes, serialisers,
                     bodyParameterIndex));
+            if (inputHandlerEngaged) {
+                handlers.add(NonBlockingHandler.INSTANCE);
+            }
+
         }
 
         // given that we may inject form params in the endpoint we need to make sure we read the body before

--- a/independent-projects/resteasy-reactive/server/runtime/src/main/java/org/jboss/resteasy/reactive/server/handlers/InputHandler.java
+++ b/independent-projects/resteasy-reactive/server/runtime/src/main/java/org/jboss/resteasy/reactive/server/handlers/InputHandler.java
@@ -25,13 +25,13 @@ import org.jboss.resteasy.reactive.server.spi.ServerRestHandler;
 public class InputHandler implements ServerRestHandler {
 
     final long maxBufferSize;
-    private volatile Executor executor;
-    private final Supplier<Executor> supplier;
+    private volatile Executor workerExecutor;
+    private final Supplier<Executor> workerExecutorSupplier;
     private final ClassLoader originalTCCL;
 
-    public InputHandler(long maxBufferSize, Supplier<Executor> supplier) {
+    public InputHandler(long maxBufferSize, Supplier<Executor> workerExecutorSupplier) {
         this.maxBufferSize = maxBufferSize;
-        this.supplier = supplier;
+        this.workerExecutorSupplier = workerExecutorSupplier;
         // capture the proper TCCL in order to avoid losing it to Vert.x in dev-mode
         this.originalTCCL = Thread.currentThread().getContextClassLoader();
 
@@ -93,8 +93,8 @@ public class InputHandler implements ServerRestHandler {
             data.add(event);
             if (dataCount > maxBufferSize) {
                 context.serverRequest().pauseRequestInput();
-                if (executor == null) {
-                    executor = supplier.get();
+                if (workerExecutor == null) {
+                    workerExecutor = workerExecutorSupplier.get();
                 }
                 //super inefficient
                 //TODO: write a stream that just uses the existing vert.x buffers
@@ -107,7 +107,7 @@ public class InputHandler implements ServerRestHandler {
                 }
                 //todo timeout
                 context.setInputStream(context.serverRequest().createInputStream(ByteBuffer.wrap(ar)));
-                context.resume(executor);
+                context.resume(workerExecutor);
             }
         }
     }

--- a/independent-projects/resteasy-reactive/server/vertx/src/test/java/org/jboss/resteasy/reactive/server/vertx/test/BodyPayloadBlockingAllowedTest.java
+++ b/independent-projects/resteasy-reactive/server/vertx/src/test/java/org/jboss/resteasy/reactive/server/vertx/test/BodyPayloadBlockingAllowedTest.java
@@ -1,0 +1,84 @@
+package org.jboss.resteasy.reactive.server.vertx.test;
+
+import static io.restassured.RestAssured.given;
+import static org.hamcrest.CoreMatchers.equalTo;
+
+import java.util.Base64;
+
+import jakarta.ws.rs.Consumes;
+import jakarta.ws.rs.POST;
+import jakarta.ws.rs.Path;
+import jakarta.ws.rs.Produces;
+import jakarta.ws.rs.core.MediaType;
+
+import org.jboss.resteasy.reactive.server.core.BlockingOperationSupport;
+import org.jboss.resteasy.reactive.server.vertx.test.framework.ResteasyReactiveUnitTest;
+import org.jboss.shrinkwrap.api.ShrinkWrap;
+import org.jboss.shrinkwrap.api.spec.JavaArchive;
+import org.junit.jupiter.api.Test;
+import org.junit.jupiter.api.extension.RegisterExtension;
+
+import io.smallrye.mutiny.Uni;
+
+public class BodyPayloadBlockingAllowedTest {
+
+    @RegisterExtension
+    static ResteasyReactiveUnitTest test = new ResteasyReactiveUnitTest()
+            .setArchiveProducer(() -> ShrinkWrap.create(JavaArchive.class)
+                    .addClasses(TestResource.class));
+
+    @Test
+    void testSmallRequestForNonBlocking() {
+        doTest(5_000, "non-blocking", false);
+    }
+
+    @Test
+    void testLargeRequestForNonBlocking() {
+        doTest(5_000_000, "non-blocking", false);
+    }
+
+    @Test
+    void testSmallRequestForBlocking() {
+        doTest(5_000, "blocking", true);
+    }
+
+    @Test
+    void testLargeRequestForBlocking() {
+        doTest(5_000_000, "blocking", true);
+    }
+
+    private static void doTest(int size, String path, boolean blockingAllowed) {
+        given() //
+                .body(String.format("{\"data\":\"%s\"}", getBase64String(size)))
+                .header("Content-Type", MediaType.TEXT_PLAIN)
+                .when().post("/test/" + path)
+                .then()
+                .statusCode(200)
+                .body(equalTo("" + blockingAllowed));
+    }
+
+    private static String getBase64String(int size) {
+        return Base64.getEncoder().encodeToString(new byte[size]);
+    }
+
+    @Path("test")
+    public static class TestResource {
+
+        @Path("non-blocking")
+        @POST
+        @Consumes(MediaType.TEXT_PLAIN)
+        @Produces(MediaType.TEXT_PLAIN)
+        public Uni<Boolean> nonBlocking(String request) {
+            return Uni.createFrom().item(BlockingOperationSupport::isBlockingAllowed);
+        }
+
+        @Path("blocking")
+        @POST
+        @Consumes(MediaType.TEXT_PLAIN)
+        @Produces(MediaType.TEXT_PLAIN)
+        public Boolean blocking(String request) {
+            return BlockingOperationSupport.isBlockingAllowed();
+        }
+
+    }
+}

